### PR TITLE
Match GPU overwritten functions with SQL functions from FunctionRegistry

### PR DESCRIPTION
**Pull request by abellina**
This change would add a dynamic table of Spark SQL functions leveraging Spark's [FunctionRegistry](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala#L192), and then tying it to the override rules in the plugin.

A different markdown page may be nice for this too, it is roughtly 300 functions. We can pull more info from the registry also, like function signatures, but I am not sure how far we want to go. The main thing is that it would change automatically and if someone wants to find whether a SQL function has potential GPU support, it is a quick lookup.

---- 

Update: Instead of a different table, we decided to take the _Expressions_ table we already had and augment it with the sql function from the `FunctionRegistry` to begin with.

### CODE REVIEW

**Comment by revans2** [comment]
I'm not sure this is going to do what we want. For example `null_if`, `nvl`, and `nvl2` are marked as not supported, but they are supported, because they are instances of `RuntimeReplaceable` and are translated into other expressions.  the problem with this approach is that `RuntimeReplaceable` requires you to actually instantiate it to know what it is going to replace it with.

**Comment by abellina** [comment]
We discussed more on this and will change it where we find the Spark SQL function(s) for expressions that we do support because of what you brought up @revans2. We also discussed trying to annotate our functions, not unlike the registry, to be able to expose things like datatype support. When we have annotations we can then bubble them up to configs.md in a similar place. This seems to me like a different issue though, so will find a good example and file that separately.

**Comment by abellina** [comment]
@jlowe @revans2 thanks for the suggestions, please take a look at this rendered version: https://github.com/NVIDIA/spark-rapids/blob/230edc57c084ed2f3d0ca9a6c4e470e4a5eac7fc/docs/configs.md, and see if this is more in line to what you were thinking.

**Comment by abellina** [comment]
I see they ended up in the anchor tag too. Let me change that.

**Comment by abellina** [comment]
Ok, sorry about that, looks better now.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala [hunk 1]
```diff
@@ -654,7 +670,16 @@ object RapidsConf {
 
       printToggleHeader("Expressions\n")
     }
-    GpuOverrides.expressions.values.toSeq.sortBy(_.tag.toString).foreach(_.confHelp(asTable))
+    val sqlFunctionsByClass = ConfHelper.getSqlFunctionsByClass
+    GpuOverrides.expressions.toSeq.sortBy(_._2.tag.toString).foreach { case (expr, rule) =>
+      val sqlFunctions = sqlFunctionsByClass.get(expr.getCanonicalName)
+      if (sqlFunctions.isDefined) {
+        val nameWithSqlFunctions = s"${rule.confKey}${sqlFunctions.get.mkString(" (", ", ", ")")}"
+        rule.confHelp(asTable, nameWithSqlFunctions)
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala:678]: I personally don't like the sql functions with the name of the config.  It feels like we are mixing concerns here.  It should either be a part of the description or it should be its own separate field.

Comment (jlowe) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala:678]: +1, I think this is interesting enough on its own to be a separate field, `SQL Function` or something similar.

Comment (abellina) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala:678]: Sounds good.

**Comment by abellina** [comment]
@kuhushukla sorry I am done with my changes :)

Here's what it looks like now: https://github.com/NVIDIA/spark-rapids/blob/e424b117c86f9cac238a880618b3979ee3f2ddc1/docs/configs.md

**Comment by revans2** [comment]
I assume that you didn't do this, but it looks like the tables for a lot of the other operators is messed up.  The description is not being printed.

**Comment by abellina** [comment]
@revans2 good eye, it was probably me. Sorry @kuhushukla

**Comment by revans2** [comment]
build

**Comment by abellina** [comment]
Ok now only the Expressions table changed. Note, I am a little worried that the rendered jekyll site will truncate it _further._ We may need to widen the pages there, or remove something from the expressions table (like the `spark.rapids` prefix). I think we want to widen the page before removing the `spark.rapids` prefix.

**Comment by revans2** [comment]
Removing the spark.rapids prefix makes the table much less usable. I cannot copy and paste any more.

**Comment by abellina** [comment]
Yeah we can follow up with jekyll style updates, different pr I'd say. https://github.com/NVIDIA/spark-rapids/issues/340

**Pull request merged**